### PR TITLE
refactor(cli): support for `.cts`、`.mts` and follows `type: module`

### PR DESCRIPTION
**Pull request by molvqingtai**
## Related issue

Enhanced configuration files, added support for `.cts` and `.mts`, and fixed the problem of not following the user `type: module`.

## summary

`cli` will automatically register `loader` for `.ts | .cts` files, for `.ts` files marked with `type: nodule` or `.mts` will be considered as ESM and need to use `cli` to specify the loader.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

rspack.config.ts:
```ts
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))

export default {
  mode: 'production',
  entry: path.resolve(__dirname, 'src/index.ts'),
  output: {
    path: path.resolve(__dirname, 'dist')
  }
}

```
package.json
```json
{
  "type": "module",
  "scripts": {
    "build": "node --loader ts-node/esm rspack build -c rspack.config.ts",
  },
  "devDependencies": {
    "@types/node": "^20.3.1",
    "ts-node": "^10.9.1",
    "typescript": "^5.1.3"
  }
}

```
tsconfig.json
```json
{
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "allowImportingTsExtensions": true,
    "noEmit": true
  }
}
```
It should work. 🎉

</details>

### CODE REVIEW

**Comment by hardfist** [comment]
we discuss how to support .ts in https://github.com/web-infra-dev/rspack/issues/2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

**Comment by molvqingtai** [comment]
> we discussed how to support .ts in #2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

I tried to use rspack self-compilation in the code, but the output file seems to have a problem and cannot be imported or required correctly.

https://github.com/molvqingtai/rspack/blob/main/packages/rspack-cli/src/utils/loadConfig.ts#L32

**Comment by hardfist** [comment]
> I tried to use rspack self-compilation in the code, but the output file seems to have a problem and cannot be imported or required correctly.
> 
use rspack to bundle config is not ideal which is disscussed in #2237

**Comment by hardfist** [comment]
it's a BUG that rspack not supports "type" property of packageJson, but it should **not** be fixed by bundle config file

**Comment by hardfist** [comment]
i think this should break into two parts type field support and mts|cts support，and we dont plan to support mts|cts

**Comment by molvqingtai** [comment]
> we discussed how to support .ts in #2237 and decide to use transform strategy other than bundle strategy, so use esbuild to prebundle config is not ideal

Hey, I have refactored the code and updated the documentation. I used [jiti](https://github.com/unjs/jiti) without bundling packages, and the dependency issue mentioned in #2237 has been resolved.

**Comment by hardfist** [comment]
> I tested jiti because it only runs the config file, so it's very fast, like node VM, unlike ts-node which pollutes the global scope.

sorry it seems webpack doesn't support `mts | cts | ts with type:module` either, and this implementation is too complicated and introduces tons of dependency(jiti uses babel) and edge case.
if you need to use `mts | cts | ts with type:module` i think you could use jiti on userland

**Comment by alexander-akait** [comment]
@hardfist 

> @alexander-akait does webpack have any plan to support mts | cts | ts with type:module config?

I thought about it, but I still think that it should be on the user side, there are a lot of different tools and approaches, but I opened to discussion

**Comment by hardfist** [comment]
https://github.com/webpack/webpack-cli/pull/3831 since webpack has supported mts|cts now, we may support it as well, @molvqingtai are you still willing to work on this?

**Comment by alexander-akait** [comment]
@hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc

**Comment by molvqingtai** [comment]
> @hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc

Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.

Use babel or other transform? I don't have a good idea at present.

**Comment by alexander-akait** [comment]
@molvqingtai 

> Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.

Why? You can't handle all possible tools and their output

**Comment by molvqingtai** [comment]
> @molvqingtai
> 
> > Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> 
> Why? You can't handle all possible tools and their output

ESM/CJS compatibility issues should not be thrown to the user. As you can imagine, when the user does not specify a loader and causes an error, it will be a bad experience to search the document for a solution.

**Comment by molvqingtai** [comment]
> > @molvqingtai
> > > Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> > 
> > 
> > Why? You can't handle all possible tools and their output
> 
> ESM/CJS compatibility issues should not be thrown to the user. As you can imagine, when the user does not specify a loader and causes an error, it will be a bad experience to search the document for a solution.



> 
Using `jiit` is currently the simplest solution while maintaining a good DX. As for edge cases, both nuxt (https://github.com/nuxt/vite/pull/201) and tailwindcss(https://github.com/tailwindlabs/tailwindcss/pull/10785) are `jiti` to load configuration files. I believe this solution is stable enough.

**Comment by alexander-akait** [comment]
@molvqingtai I strongly disagree, but I will not argue, we already tried to do it and faced with a lot of problems when new things appear (syntax/different outputs/just features like swc for `ts-node`/ES modules), why I should use babel (and dependecies) when I want to use swc. Setting up the runtime environment is just the task of the developer and no one else, you can use `jiti` under the hood if **nothing** was configured (or no options/setting provided), but always using and not allowing customization is a pretty bad idea, you can of course not take someone else's experience and go through the same problems again, your choice

**Comment by alexander-akait** [comment]
And yes, don't forget about performance, babel is not fast

**Comment by molvqingtai** [comment]
> And yes, don't forget about performance, babel is not fast


Of course, you can use swc. https://github.com/tailwindlabs/tailwindcss/pull/10785/commits/f1614c8d9ef50d507e271a30de6213cbaa932f0f

**Comment by hardfist** [comment]
> > @hardfist note - we don't use `jiti`, we allow to use any tools i.e. `ts-node`, `babel` and etc
> 
> Let the user specify node loader like webpack [L11](https://github.com/webpack/webpack-cli/blob/bfddce332f91214b55cde5389c7920a2b7f6f576/test/build/config-format/typescript-esnext-mjs/typescript.test.js#L11), I don't think this is a good idea.
> 
> Use babel or other transform? I don't have a good idea at present.

since esm is very tricky and contains tons of edge case，i dont think let advanced user to handle esm-loader itself is a bad idea，@xiaoxiangmoe mentions jiti has some tricky design and may contain some edge case

**Comment by hardfist** [comment]
i do agree jiti is simple and can handle 80 percent cases，but jiti has  many options which we cant expose to users and i do notice it has some lockin design which will break when javascript has new syntax or nodes invents new esm only api

**Comment by molvqingtai** [comment]
> i do agree jiti is simple and can handle 80 percent cases，but jiti has many options which we cant expose to users and i do notice it has some lockin design which will break when javascript has new syntax or nodejs invents new esm only api

I agree with you. Can we retain the ability to customize loader and use jiit to load when no loader is specified? This ensures that the ESM configuration file can be used out of the box, and when jiti does not follow the new syntax, or some other edge cases, users can customize the loader processing.

**Comment by xiaoxiangmoe** [comment]
I don't think using jiti is a good idea because it has some edge case. For example:

### TypeScript commonjs syntax

```ts
// file: rspack.config.ts
import * as path from "path";

const config = {
	mode: "production",
	entry: "./main.ts",
	output: {
		path: path.resolve(__dirname, "dist"),
		filename: "foo.bundle.js"
	}
};

export = config;// this stand for module.exports = config in commonjs
```

This is the usage that rspack provided now. But jiti can't handle it correctly. This will break users.

### Abusing cjs and esm

import.meta.url exist in cts
```ts
console.log(import.meta.url) // file:///Users/me/ts-node-playground/a.cts
```
__filename exist in mts

```ts
console.log(__filename) // /Users/me/ts-node-playground/a.mts
```

If we provide these feature. Should we also provide this in `.js`, `.cjs` and `.mjs`?

**Comment by hardfist** [comment]
> To summarize the issues listed, in order to be consistent with webpack (important), if there is no better suggestion, I will delete jiit and adopt webpack's solution.

thank you for your understanding, the current wepback's way is a right(contains less edge case) but not convenient way and we may iterate to improve the DX of this in the future

**Comment by hardfist** [comment]
Sometimes`` it seems to me that the most ideal for configurations was pure JS with jsdocs comments for types - no extra loaders, no extra transpilation, no extra overhead.

i prefer this way too，and that is why all of our examples written in this way

**Comment by molvqingtai** [comment]
@hardfist @alexander-akait 
The refactoring is complete and the description has been updated

File: packages/rspack-cli/package.json [hunk 1]
```diff
@@ -26,27 +26,23 @@
...
-  },
   "dependencies": {
     "@discoveryjs/json-ext": "^0.5.7",
     "@rspack/core": "workspace:*",
     "@rspack/dev-server": "workspace:*",
     "colorette": "2.0.19",
+    "interpret": "^3.1.1",
+    "rechoir": "^0.8.0",
     "semver": "6.3.0",
+    "type-fest": "^2.14.0",
```
Comment (hardfist) [inline packages/rspack-cli/package.json:45]: this dependency is unnecessary we can write our own type

File: packages/rspack-cli/src/utils/crossImport.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { pathToFileURL } from "url";
+import isEsmFile from "./isEsmFile";
+
+const crossImport = async <T = any>(
+	path: string,
+	cwd = process.cwd()
+): Promise<T> => {
+	if (isEsmFile(path, cwd)) {
+		const url = pathToFileURL(path).href;
+		const { default: config } = await import(url);
```
Comment (hardfist) [inline packages/rspack-cli/src/utils/crossImport.ts:10]: this will crash node in jest test due to https://github.com/nodejs/node/issues/35889, and I'm not sure how webpack-cli workaround it, @alexander-akait can help me here?

**Comment by hardfist** [comment]
@molvqingtai thank you for your hard work and detailed test, the overall work LGTM

File: packages/rspack-cli/package.json (hunk 1 above)
Comment (molvqingtai) [reply packages/rspack-cli/package.json:45]: > this dependency is unnecessary we can write our own type

Looking at the type-fest source code, packageJson actually has a large number of type definitions. Is it necessary to copy the complete type definition, or simply write it as {"type": "module"}

File: packages/rspack-cli/src/utils/crossImport.ts (hunk 2 above)
Comment (alexander-akait) [reply packages/rspack-cli/src/utils/crossImport.ts]: We spawn a new process for the each test in CLI integration tests, so we don't have `jest` in scope, to collect coverage we use `nyc` (yeah, it is deprecated, but works).

Comment (hardfist) [reply packages/rspack-cli/src/utils/crossImport.ts]: i dont get it why nyc can workaround this problem here since it still runs jest  and it seems it doesnt enable vm-modules，how can jest run dynamic import without vm-moduels enabled？

Comment (alexander-akait) [reply packages/rspack-cli/src/utils/crossImport.ts]: `nyc` is not a workaround, the main workaround - we spawn a new process inside `it`, i.e. (pseudo code):
```
it("serve command", () => { 
  spawnNewProcess(__dirname, ["serve"]); 
});
```

So we don't have `jest` inside a new spawed process, it also allows you to test as if the user has run it - no polluting, just a fresh process.

`nyc` is a workaround to get coverage after testing, because we don't have `jest` inside a new spawned process we can't collect lines, so here we use `nyc` inside a new spawned process and collect coverage (but we collect it only for CI)

File: packages/rspack-cli/tests/build/config/moonrepo/rspack.config.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
-// @ts-ignore
+// @ts-ignore: Because dynamically create {"type": "module"}
```
Comment (hardfist) [inline packages/rspack-cli/tests/build/config/moonrepo/rspack.config.ts:1]: you can commit node_modules to git for test

File: packages/rspack-cli/package.json (hunk 1 above)
Comment (hardfist) [reply packages/rspack-cli/package.json:45]: write type: module is enough for our case

**Comment by hardfist** [comment]
@molvqingtai  I'll keep fixing the left issue to get this pr merged ASAP if you don't mind

**Comment by hardfist** [comment]
closes #3609

**Comment by molvqingtai** [comment]
> @molvqingtai I'll keep fixing the left issue to get this pr merged ASAP if you don't mind

Of course

**Comment by hardfist** [comment]
!canary

**Review by hyf0 — APPROVED** [review]

**Comment by hardfist** [comment]
@molvqingtai thank you for your great work
